pallets/asset/src/lib.rs

```diff
@@ -2269,4 +1888,498 @@ impl<T: Trait> Module<T> {
+                &extension_details.extension_type,
+                &extension_details.extension_id
+            ),
+            Error::<T>::IncompatibleExtensionVersion
+        );
+        // Ensure the hard limit on the count of maximum transfer manager an asset can have.
+        Self::ensure_max_limit_for_tm_extension(&extension_details.extension_type, &ticker)?;
+
+        // Update the storage
+        <ExtensionDetails<T>>::insert(
+            (ticker, &extension_details.extension_id),
+            extension_details.clone(),
+        );
+        <Extensions<T>>::append(
+            (ticker, &extension_details.extension_type),
+            extension_details.extension_id.clone(),
+        );
+
+        let event = RawEvent::ExtensionAdded(
+            my_did,
+            ticker,
+            extension_details.extension_id,
+            extension_details.extension_name,
+            extension_details.extension_type,
+        );
+        Self::deposit_event(event);
+        Ok(())
+    }
+
+    fn base_remove_smart_extension(
+        origin: T::Origin,
+        ticker: Ticker,
+        extension_id: T::AccountId,
+    ) -> DispatchResult {
+        // Ensure the extrinsic is signed and have valid extension id.
+        let did = Self::ensure_signed_and_validate_extension_id(origin, &ticker, &extension_id)?;
+        let extension_detail_key = (ticker, extension_id.clone());
+        ensure!(
+            <ExtensionDetails<T>>::contains_key(&extension_detail_key),
+            Error::<T>::NoSuchSmartExtension
+        );
+
+        let extension_type = Self::extension_details(&extension_detail_key).extension_type;
+
+        // Remove the storage reference for the given extension_id.
+        // The order of SEs do not matter, so `swap_remove` is OK.
+        <Extensions<T>>::mutate(&(ticker, extension_type), |extension_list| {
+            if let Some(pos) = extension_list.iter().position(|ext| ext == &extension_id) {
+                extension_list.swap_remove(pos);
+            }
+        });
+        <ExtensionDetails<T>>::remove(extension_detail_key);
+
+        Self::deposit_event(RawEvent::ExtensionRemoved(did, ticker, extension_id));
+        Ok(())
+    }
+
+    fn set_archive_on_extension(
+        origin: T::Origin,
+        ticker: Ticker,
+        extension_id: T::AccountId,
+        archive: bool,
+    ) -> DispatchResult {
+        // Ensure the extrinsic is signed and have valid extension id.
+        let did = Self::ensure_signed_and_validate_extension_id(origin, &ticker, &extension_id)?;
+
+        // Mutate the extension details
+        <ExtensionDetails<T>>::try_mutate((ticker, &extension_id), |details| {
+            let error = |is_archive: bool| {
+                is_archive
+                    .then_some(Error::<T>::AlreadyArchived)
+                    .unwrap_or(Error::<T>::AlreadyUnArchived)
+            };
+
+            ensure!(details.is_archive != archive, error(archive));
```

Comment:
Closure is used once?

(It's hard to tell in this diff what are changes to the existing code and what are just pure move-arounds of code... For the future, I would have preferred to split that into 2 separate PRs)